src/shared/services/shortcut-action.service.ts

```diff
@@ -0,0 +1,200 @@
+/*
+ * This file is part of ng2-json-editor.
+ * Copyright (C) 2016 CERN.
+ *
+ * ng2-json-editor is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * ng2-json-editor is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with ng2-json-editor; if not, write to the Free Software Foundation, Inc.,
+ * 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
+ * In applying this license, CERN does not
+ * waive the privileges and immunities granted to it by virtue of its status
+ * as an Intergovernmental Organization or submit itself to any jurisdiction.
+ */
+
+import { Injectable } from '@angular/core';
+import { EmptyValueService } from './empty-value.service';
+import { JsonStoreService } from './json-store.service';
+import { JsonSchemaService } from './json-schema.service';
+import { DomUtilService } from './dom-util.service';
+import { TabIndexService } from './tab-index.service';
+import { PathUtilService } from './path-util.service';
+import { List } from 'immutable';
+
+@Injectable()
+export class ShortcutActionService {
+
+  constructor(public emptyValueService: EmptyValueService,
+              public domUtilService: DomUtilService,
+              public jsonStoreService: JsonStoreService,
+              public jsonSchemaService: JsonSchemaService,
+              public pathUtilService: PathUtilService,
+              public tabIndexService: TabIndexService) { }
+
+  addRootAction(path: Array<any>): void {
+    this.addAction(path, true);
+  }
+
+  addParentAction(path: Array<any>): void {
+    this.addAction(path, false);
+  }
+
+  private addAction(path: Array<any>, root: boolean): void {
+    let _path = this.pathUtilService.getNearestOrRootArrayParentInPath(path, root);
+    this.addNewElementInArray(_path, this.jsonSchemaService.getSchemaFromPath(_path));
+  }
+
+  /**
+   * @param {Array<any>} path - Path of the array parent that the element is about to be inserted
+   * @param {Object} schema - Schema of the element that is about to be inserted
+   */
+  private addNewElementInArray(path: Array<any>, schema: Object) {
+    let itemSchema = schema['items'];
+    let emptyValue = this.emptyValueService.generateEmptyValue(itemSchema);
+    let values = this.jsonStoreService.getIn(path) || List.of([]);
```

Comment:
`List()` instead of `List.of([])`

Also is it possible to reuse `AbstractListField`'s add new element functionality here.